DOCSP-30550: GridFS guide (#89)
mongodb/docs-rust: source/fundamentals/gridfs.txt

.. _rust-gridfs:

======
GridFS
======

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: large files, storage, code example

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to store and retrieve large files in

> MongoDB using **GridFS**. GridFS is a specification that describes how

Reviewer: ```suggestion
MongoDB by using **GridFS**. GridFS is a specification that describes how
```

to split files into chunks during storage and reassemble them during retrieval.
The {+driver-short+} implementation of GridFS manages the operations and
organization of the file storage.

Use GridFS if the size of your file exceeds the BSON document size limit of
16 MB. GridFS also helps you access files without loading the entire file into

> memory. For more detailed information on whether GridFS is suitable for your

Reviewer: ```suggestion
memory. For more detailed information about whether GridFS is suitable for your
```

use case, see the :manual:`GridFS </core/gridfs>` page in the Server manual.

This guide includes the following sections:

- :ref:`How GridFS Works <rust-gridfs-info>` describes how files are organized and stored
  according to GridFS specifications

- :ref:`Create a GridFS Bucket <rust-gridfs-create>` describes how to create a GridFS bucket
  and configure options 

- :ref:`Upload Files <rust-gridfs-upload>` describes how to upload a file to a
  GridFS bucket by using an upload stream

- :ref:`Download Files <rust-gridfs-download>` describes how to download a file from
  a GridFS bucket by using a download stream

- :ref:`Retrieve File Information <rust-gridfs-retrieve>` describes how to access
  information about files in a GridFS bucket 

- :ref:`Rename Files <rust-gridfs-rename>` describes how to rename a file in a GridFS bucket
  
- :ref:`Delete Files <rust-gridfs-delete>` describes how to delete files from a GridFS bucket
  
- :ref:`Delete a GridFS Bucket <rust-gridfs-drop>` describes how to delete a GridFS bucket

- :ref:`Additional Information <rust-gridfs-addtl-info>` provides links to resources and API documentation

>   for types and methods mentioned in this guide

Reviewer: The titles in the TOC are descriptive enough to help a user find what they need, so good job there!

.. _rust-gridfs-info:

How GridFS Works
----------------

> GridFS organizes files in a **bucket**, a group of MongoDB collections

Reviewer: Either one seems fine to me; the suggestion is wordier, but probably clearer. The SG doesn't specifically discuss this case or clarifying non-restrictive clauses.

> GridFS organizes files in a **bucket**, a group of MongoDB collections

Author: I added "which is" for clarity, but reworded slightly to make it less wordy

that contain the chunks of files and descriptive information. Buckets
contain the following collections, named according to the convention
defined in the GridFS specification:

- ``chunks``, which stores the binary file chunks

> - ``files``, which stores the file metadata

Reviewer: this got me too--the singular collection "chunks" has a plural name, so i think the original is right

> When you create a new GridFS bucket, the {+driver-short+} creates the ``chunks``

Author: Based on the other drivers' GridFS pages and the Server docs, my understanding is that the driver does this work

and ``files`` collections, prefixed with the default bucket name ``fs``, unless
you specify a different name. The driver also creates an index on each

> collection to ensure efficient retrieval of files and related

Reviewer: I think a list is fine here, but SG suggests [not introducing lists with fragments.](https://www.mongodb.com/docs/meta/style-guide/style/lists/#:~:text=Avoid%20using%20fragments%20to%20introduce%20lists.%20Fragments%20can%20be%20harder%20to%20understand%20than%20sentences.) List items should also be capitalized.

metadata. The driver creates the GridFS bucket on the first write
operation only if it does not already exist. The driver creates indexes only if

> they do not exist and when the bucket is empty. For more information on

Author: I reworded, hopefully it's more clear now

GridFS indexes, see the :manual:`GridFS Indexes </core/gridfs/#gridfs-indexes>`
page in the Server manual.

When storing files in a GridFS bucket, the {+driver-short+} splits the files into
smaller pieces, each represented by a separate document in the ``chunks`` collection.
The driver also creates a document in the ``files`` collection that contains
a unique file ID, file name, and other file metadata. You can upload the file from
from an upload stream. The following diagram describes how GridFS splits

> files when uploading to a bucket:

Reviewer: If you want to keep the list, I'd probably keep the bit about dividing the files into smaller pieces. Should also capitalize the list items and reword the intro so it [doesn't end with "the following".](https://www.mongodb.com/docs/meta/style-guide/style/lists/#:~:text=Don%E2%80%99t%20tell%20users%20to%20%22do%20the%20following.%22%20The%20verb%20do%20is%20weak%2C%20using%20following%20as%20a%20noun%20in%20this%20context%20is%20incorrect%2C%20and%20the%20whole%20phrase%20is%20ambiguous.)

.. figure:: /includes/figures/GridFS-upload.png
   :alt: A diagram that shows how GridFS uploads a file to a bucket

When retrieving files, GridFS fetches the metadata from the ``files``
collection in the specified bucket and uses the information to reconstruct
the file from documents in the ``chunks`` collection. You can read the file
into memory or output it to a stream.

.. _rust-gridfs-create:

> Create a GridFS Bucket

Reviewer: ^^^ worth clarifying for sure

> Create a GridFS Bucket

Author: Really good point, I did some experimenting to make sure but yes, the bucket is only created upon the first write operation. Changed!

----------------------

> To begin storing files in a GridFS bucket, create a bucket or get a reference to
> an existing one. 
> 
> The following example calls the ``gridfs_bucket()`` method on a database to create
> a new GridFS bucket:

Reviewer 2: ```suggestion
To upload new files or access existing ones, create a reference to a GridFS bucket.

The following example calls the ``gridfs_bucket()`` method on a database to create a reference to a either a new GridFS bucket or an existing one:
```

Reviewer 3: Suggestion:
I think the first sentence should provide more information about the subject ("how to upload files"). While "create a reference" is a necessary step, it doesn't describe how to upload. Perhaps separating out "upload new files" and "access existing [files]" could be separated into two sentences so that each sentence could explain how to accomplish the complete task. E.g.

"To access an existing bucket, call the gridfs_bucket() on the database that contains the GridFS bucket. To upload a file to that bucket, call the ... method on the instance of the bucket."

Author: Since the following section ("Upload Files") explains how to upload a file, I don't think that info is needed here. I did reword this part of the "Reference a GridFS Bucket" section though to make it more clear.

.. literalinclude:: /includes/fundamentals/code-snippets/gridfs.rs
   :start-after: start-create
   :end-before: end-create
   :language: rust
   :copyable:
   :dedent:

> You can specify a custom bucket name and other options in a ``GridFsBucketOptions``

Reviewer 2: Is there a default name?

Author: Yes the default name is "fs" -- I included that in the table, do you think I should mention outside the table as well?

Reviewer 2: I missed that, I think the table is sufficient.

.. list-table::
   :widths: 1 1 2
   :stub-columns: 1
   :header-rows: 1

   * - Method
     - Possible Values 
     - Description

   * - | ``bucket_name()``
     - | Any ``String`` value
     - | Specifies a bucket name, which is set to ``fs`` by default
       
   * - ``chunk_size_bytes()``
     - Any ``u32`` value
     - Specifies the chunk size, which is 255 KiB by default, used to break the file into chunks

   * - | ``write_concern()``
     - | ``WriteConcern::w``,
       | ``WriteConcern::w_timeout``,
       | ``WriteConcern::journal``
     - | Specifies the bucket's write concern, which is set to the database's write concern by default

   * - ``read_concern()``
     - ``ReadConcernLevel::Local``,
       ``ReadConcernLevel::Majority``,
       ``ReadConcernLevel::Linearizable``,
       ``ReadConcernLevel::Available``,
       ``ReadConcernLevel::Snapshot``
     - Specifies the bucket's read concern, which is set to the database's read concern by default

   * - | ``selection_criteria()``
     - | ``SelectionCriteria::ReadPreference``,
       | ``SelectionCriteria::Predicate``
     - | Specifies whether the driver compares numeric strings as numbers

> The following example specifies a ``GridFsBucketOptions`` instance and creates a

Reviewer 2: Again, because the creation doesn't happen until the first write operation:
```suggestion
The following example specifies a ``GridFsBucketOptions`` instance and references a
```

GridFS bucket with a custom name:

.. literalinclude:: /includes/fundamentals/code-snippets/gridfs.rs
   :start-after: start-create-opts
   :end-before: end-create-opts
   :language: rust
   :copyable:
   :dedent:

.. _rust-gridfs-upload:

Upload Files
------------

You can upload a file to a GridFS bucket by opening an upload stream and 
writing your file to the stream. Call the ``open_upload_stream()`` method on
your bucket instance to open the stream. Then, call the ``write_all()`` method

> with your file bytes as a parameter to upload the file.

Reviewer: ```suggestion
and pass your file bytes as a parameter to upload the file.
```

The following example uses an upload stream to upload a file called ``"example.txt"``
to a GridFS bucket:

.. literalinclude:: /includes/fundamentals/code-snippets/gridfs.rs
   :start-after: start-upload
   :end-before: end-upload
   :language: rust
   :copyable:
   :dedent:

.. _rust-gridfs-download:

Download Files
--------------

You can download a file from a GridFS bucket by opening a download stream and
reading from the stream. Call the ``open_download_stream()`` method on
your bucket instance, specifying the desired file's ``_id`` value as a parameter.

> Then, call the ``read_to_end()`` method with a vector parameter to read and store the
> file bytes.

Reviewer: ```suggestion
Then, call the ``read_to_end()`` method and pass a vector parameter to read and store the
file bytes.
```

will a user know what a vector parameter is?

> The following example uses an download stream to download a file with an ``_id`` value

Reviewer: ```suggestion
The following example uses a download stream to download a file with an ``_id`` value
```

- ``_id``: the file ID
- ``length``: the file size
- ``chunk_size_bytes``: the size of the file's chunks
- ``upload_date``: the file's upload date and time
- ``filename``: the name of the file
- ``metadata``: a document that stores user-specified metadata

Call the ``find()`` method on a GridFS bucket instance to retrieve
files from the bucket. The method returns a cursor instance
from which you can access the results.

The following example retrieves and prints the length of each file in a GridFS bucket:

.. literalinclude:: /includes/fundamentals/code-snippets/gridfs.rs
   :start-after: start-retrieve
   :end-before: end-retrieve
   :language: rust
   :copyable:
   :dedent:

.. _rust-gridfs-rename:

Rename Files
------------

You can update the name of a GridFS file in your bucket by calling the ``rename()`` method
on a bucket instance. Pass the target file's ``_id`` value and the new file name as
parameters to the ``rename()`` method.

.. note::

   The ``rename()`` method only supports updating the name of one file at
   a time. To rename multiple files, retrieve a list of files matching the
   file name from the bucket, extract the ``_id`` field from the files you

>    want to rename, and pass each value in separate calls to the ``rename()``

Reviewer: ```suggestion
   want to rename, and pass each ``_id`` value in separate calls to the ``rename()``
```

>    method.

Reviewer: makes sense

>    method.

Author: Added an example! I think it might be a little repetitive, since the example isn't much different from the preceding one other than the number of times `rename()` is used, but let me know what you think

The following example updates the ``filename`` field of the file containing an ``_id`` value
of ``3289`` to ``"new_file_name"``:

.. literalinclude:: /includes/fundamentals/code-snippets/gridfs.rs
   :start-after: start-rename
   :end-before: end-rename
   :language: rust
   :copyable:
   :dedent:

.. _rust-gridfs-delete:

Delete Files
------------

You can use the ``delete()`` method to remove a file from your bucket. To remove a
file, call ``delete()`` on your bucket instance and pass the file's ``_id`` value
as a parameter.

.. note::

   The ``delete()`` method only supports deleting one file at a time. To
   delete multiple files, retrieve the files from the bucket, extract

>    the ``_id`` field from the files you want to delete, and pass each value

Reviewer: ```suggestion
   the ``_id`` field from the files you want to delete, and pass each ``_id`` value
```

   in separate calls to the ``delete()`` method.

> The following example deletes the file in which the ``_id`` value is ``3289``:

Reviewer: ```suggestion
The following example deletes the file in which the value of the ``_id`` field is ``3289``:
```
